### PR TITLE
[BEAM-2728] Sketching most frequent

### PR DESCRIPTION
**Pull request by ArnaudFnr**
This is the last PR for [BEAM-2728] issue.
The last sketch uses AddThis' StreamSummary sketch implementation, which presents many defects for distributed processing (such as loss of information during merging...). 
That's why I will try to update this sketch as soon as possible with a better implementation.

The MostFrequent class contains several components, so I think it will be clearer to present them here:
1. Default PTransforms that are AutoValue classes like for previous sektches
2. A parent Combine class MostFrequentFn with two childs **SerializableElements** and **NonSerializableElements** to distinguish the two cases, because StreamSummary needs to receive serializable elements
3. An **ElementWrapper** in order to wrap non serializable elements into a serializable one using the Coder
4. A **StreamSummarySketch** which inherits from StreamSummary in order to implement the "equals()" method.
5. **RetrieveTopK** utility class which Contains DoFn for retrieving the top k most frequent elements from a PCollection of StreamSummarySketches. It avoid users to pull stream-lib dependencies for retrieving the top K elements.
------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand:
   - [ ] What the pull request does
   - [ ] Why it does it
   - [ ] How it does it
   - [ ] Why this approach
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

**Comment by ArnaudFnr** [comment]
R : @iemejia @jbonofre @jkff @lukecwik

File: sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java [hunk 1]
```diff
@@ -0,0 +1,747 @@
...
+ *
+ * <p>The Stream Summary uses a doubly linked-list of buckets in order to order elements
+ * by frequency. Each bucket is associated with a list of Counters, each of them storing
+ * information about one of the tracked elements : value, count and maximum potential error.
+ * <br>
+ *   An element is guaranteed to be in the top K most frequent if its guaranteed number
+ *   of hits, i.e. the {@code count minus error}, is greater than the count of the element at
+ *   the position k+1.
+ * <br>
+ *   <b>WARNING</b>: Information about the error is lost during the parallelization,
```
Comment (jkff) [inline sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java]: Does this mean that there are no guarantees whatsoever about the output of this algorithm, if merging of accumulators ever happens? I think that means that, until this issue is fixed in stream-lib, we need to not use Combine, and use a perhaps slower implementation but one that provides some guarantees.

File: sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java [hunk 2]
```diff
@@ -0,0 +1,747 @@
...
+ *
+ * <ul>
+ *   <li>The top {@code k} most frequent elements you want to output from the stream.<br>
+ *     By default, {@code k} is set to {@code 100}.
+ *   <li>The {@code capacity} of the Stream Summary sketch. It corresponds to
+ *   the maximum number of elements that can be tracked at once.
+ *   When the maximum {@code capacity} is reached, the least frequent element
+ *   will be replaced by the next untracked element.<br>
+ *     By default, the {@code capacity} is set to {@code 10000} but this value depends
+ *     highly on the use case and more specifically on the order of incoming elements.
```
Comment (jkff) [inline sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java]: I'm not sure what I'd do with this information as a user - especially considering that PCollections don't have ordering.

File: sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java [hunk 3]
```diff
@@ -0,0 +1,747 @@
...
+ * </ul>
+ *
+ *  <p><b>WARNING</b>: Keep in mind that he {@code capacity} should always be
+ *  significantly higher than {@code k}. One must ask himself the following question:<br>
```
Comment (jkff) [inline sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java]: Recommend using gender-neutral language e.g. "oneself": many users and contributors are not male.

File: sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java [hunk 4]
```diff
@@ -0,0 +1,747 @@
...
+ *    What is the lowest rank {@code x} such that any element arriving at a rank {@code y > x}
```
Comment (jkff) [inline sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java]: What does "arriving at a rank y" mean, considering lack of ordering?

File: sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java [hunk 5]
```diff
@@ -0,0 +1,747 @@
...
+      return toBuilder().setCapacity(c).build();
+    }
+
+    @Override
+    public PCollection<KV<InputT, Long>> expand(PCollection<InputT> input) {
+      final Coder<InputT> inputCoder = input.getCoder();
+      Class clazz = inputCoder.getEncodedTypeDescriptor().getRawType();
+      PCollection<KV<InputT, Long>> result;
+
+      if (Serializable.class.isAssignableFrom(clazz)) {
```
Comment (jkff) [inline sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java]: Why is this needed? And why do we ignore the input coder if the class is serializable?

File: sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java [hunk 6]
```diff
@@ -0,0 +1,747 @@
...
+     * @param c the capacity of the summary
+     */
+    public GlobalSummary<InputT> withCapacity(int c) {
+      return toBuilder().setCapacity(c).build();
+    }
+
+    @Override
+    public PCollection<KV<InputT, Long>> expand(PCollection<InputT> input) {
+      final Coder<InputT> inputCoder = input.getCoder();
+      Class clazz = inputCoder.getEncodedTypeDescriptor().getRawType();
```
Comment (jkff) [inline sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java]: This might fail given sufficient type erasure.

File: sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java [hunk 7]
```diff
@@ -0,0 +1,747 @@
...
+
+    public StreamSummarySketch() {
+      super();
+    }
+
+    public StreamSummarySketch(int capacity) {
+      super(capacity);
+    }
+
+    /** Merge several {@link StreamSummarySketch} into this one. */
```
Comment (jkff) [inline sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java]: Did you mention above that this function is broken in some way? Add a comment here?

File: sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java [hunk 8]
```diff
@@ -0,0 +1,747 @@
...
+        return false;
+      }
+      StreamSummarySketch that = (StreamSummarySketch) o;
+      return this.toString().equals(that.toString());
+    }
+  }
+
+  /**
+   * This class is used to wrap a non serializable type into a serializable wrapper which
+   * uses the {@link Coder} as serializer. This wrapper is automatically used when calling
```
Comment (jkff) [inline sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java]: Why do we need this wrapper? If it's for the sake of making the StreamSummary itself encodable, I think you could get around that by encoding it using its .peek(capacity) and decoding symmetrically.

File: sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java (hunk 1 above)
Comment (ArnaudFnr) [reply sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java]: There's indeed no guarantee on the error.. But depending on the use case this should not be as awful as it looks...

For instance, the whole algorithm (and not just the implementation) is a problem for global windows because once a capacity is reached, it seems unlikely that an untracked elements goes all the way to the top and even in the case it would, there would be a high error.

However in the context of short timing windows, like hours or maybe days depending on the volume we can support, top elements if they appear soon enough in the process will be directly on top and the error will usually be 0 because they will never be discarded.

So yeah the output depends on the order of the element, but in our case I should rather say it depends on the timeliness more than the "exact" order

In all cases I already found some resources which present Algorithms adapted to parallélisation, so I will investigate on that soon.

File: sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java (hunk 2 above)
Comment (ArnaudFnr) [reply sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java]: You're right, I should rather talk about timeliness in beam context I think.

File: sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java (hunk 3 above)
Comment (ArnaudFnr) [reply sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java]: Oh my bad of course, I was not sure if we could say "oneself"

File: sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java (hunk 4 above)
Comment (ArnaudFnr) [reply sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java]: What I want to say is :
What is the minimal number of different elements N such that in the context of the time window of a processing, every new element will have a very low probability to be considered as a frequent element at the end of the time window.

Is it more clear from a beam context?

File: sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java (hunk 6 above)
Comment (ArnaudFnr) [reply sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java]: What do you mean by sufficient type erasure?

File: sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java (hunk 5 above)
Comment (ArnaudFnr) [reply sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java]: I wanted to distinguish the cases where elements are serializable from cases where elements are not serializable, because in the latter case I'll have to wrap elements in a serializable wrapper.
This is due to the serialization of StreamSummary which serializes the tracked elements using the default method "outSream.writeObject()"
I cannot change the way the StreamSummary is serialized because of private some members.

In the first case elements don't need to be wrapped therefore it would be a loss of time to wrap them, that's why I made a distinction. 

Another solution would be to encode elements into bytes before offering them to the structure (like we did for Hyperloglog for instance) but I think this will be slower

File: sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java (hunk 7 above)
Comment (ArnaudFnr) [reply sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java]: I mentioned it in the outter class Javadoc but I can put it here two, yes.

File: sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java (hunk 8 above)
Comment (ArnaudFnr) [reply sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java]: It's not to make it encodage but to make it consistent with equals(), due to the Illegal mutation error I got

File: sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java (hunk 1 above)
Comment (jkff) [reply sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java]: I looked up the algorithm and it says that it is guaranteed to find all items with count > n/k. Is this not the case in the current implementation? If so, is that because of merging or because of something else? If it's because of merging, let's get rid of merging - that will require not using a combine fn.

File: sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java (hunk 4 above)
Comment (jkff) [reply sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java]: Beam does not have the concept of arrival order of elements, period. A runner is technically allowed to accumulate stuff and then call your DoFn's on it in random order; and if you're using one of the runners or runner modes primarily designed for batch use cases, this is pretty much what it will be like.

So, in other words, we can only give guarantees that can be justified by the Beam model itself, rather than by assumptions about run-time execution.

File: sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java (hunk 6 above)
Comment (jkff) [reply sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java]: Disregard this, I'm not sure what I meant anymore.

File: sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java (hunk 5 above)
Comment (jkff) [reply sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java]: My objection is to this:

> I cannot change the way the StreamSummary is serialized because of some private members.

I believe you don't have to serialize StreamSummary directly. Instead, you can develop a Coder for it that works like this:

encode(): call topK(Integer.MAX_VALUE), obtain a list of element/count pairs, serialize those (using the user's provided input coder)
decode(): get the element/count pairs, insert each of them into a new empty StreamSummary, return that.

File: sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java (hunk 8 above)
Comment (jkff) [reply sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java]: This you can circumvent by creating a wrapper around StreamSummary with a properly defined equals() / hashCode().

Comment (ArnaudFnr) [reply sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java]: Whoops I thought you were talking about the StreamSummarySketch wrapper.

This wrapper is used for non serizaliable elements because StreamSummary calls writeObject() for tracked elements therefore when elements are not serializable this is a problem.

But this problem will be solved if we encode the StreamSummary as you said above.

File: sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java (hunk 5 above)
Comment (ArnaudFnr) [reply sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java]: You're right, I also thought about this but don't you think this will drastically slow down the Combine processing ?

File: sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java (hunk 4 above)
Comment (ArnaudFnr) [reply sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java]: Ok I understand.

File: sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java (hunk 1 above)
Comment (ArnaudFnr) [reply sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java]: Indeed it guaranteed to find all items with count > ØN, Ø being the support between 0 and 1. 
But this gives no guarantee about the output top k elements, which are of interest here. 
Thus the guarantee on the top k depends on the estimated error e, but due to the inconsistent merging this is not possible to compute.

**Comment by robertwb** [comment]
@ArnaudFnr Any update on this?

**Comment by iemejia** [comment]
@ArnaudFnr Any updates? I have the impression there were not many things missing, are they many ? Do you intend to finish this one (even if it is not right now) ?

**Comment by jkff** [comment]
This PR is stale since it hasn't had contributor activity since March, so I'm going to close it. That said, this is a very valuable contribution and someone should pick it up from here (would happily do so myself if I had more time). I think very little remains to finish this up.
